packages/java/parser-jvm-core/src/main/java/com/vaadin/fusion/parser/core/Parser.java

```diff
@@ -68,48 +71,56 @@ public SharedStorage getStorage() {
         return storage;
     }
 
-    private class EntitiesCollector {
-        private final Set<RelativeClassInfo> endpoints;
-        private final Set<RelativeClassInfo> entities;
+    private static class ElementsCollector {
+        private final ScanResult result;
+        private final Logger logger;
 
-        public EntitiesCollector(ScanResult result, Logger logger) {
-            var endpointAnnotationName = config.getEndpointAnnotationName();
+        public ElementsCollector(ScanResult result, Logger logger) {
+            this.result = result;
+            this.logger = logger;
+        }
 
+        public Collection<RelativeClassInfo> collectEndpoints(
+                String endpointAnnotationName) {
             logger.debug(
                     "Collecting project endpoints with the endpoint annotation: "
                             + endpointAnnotationName);
 
-            endpoints = result.getClassesWithAnnotation(endpointAnnotationName)
-                    .stream().map(RelativeClassInfo::new)
+            var endpoints = result
+                    .getClassesWithAnnotation(endpointAnnotationName).stream()
+                    .map(RelativeClassInfo::new)
                     .collect(Collectors.toCollection(LinkedHashSet::new));
 
             logger.debug("Collected project endpoints: " + endpoints.stream()
                     .map(RelativeClassInfo::get).map(ClassInfo::getName)
                     .collect(Collectors.joining(", ")));
 
-            entities = endpoints.stream().flatMap(
+            return endpoints;
+        }
+
+        public Collection<RelativeClassInfo> collectEntities(
+                Collection<RelativeClassInfo> endpoints) {
+            var entities = endpoints.stream().flatMap(
                     cls -> cls.getInheritanceChain().getDependenciesStream())
-                    .collect(Collectors.toCollection(LinkedHashSet::new));
+                    .collect(Collectors.toCollection(ArrayList::new));
 
             // ATTENTION: This loop mutates the collection during processing!
```

Comment:
This doesn't look good. I think what you can do is to create a copy of the entities, iterate through the copy, and modify the entities. e.g.,

```
var originalEntities = new ArrayList<>(entities);
for (var entity : originalEntities) {
    entity.getDependenciesStream().filter(e -> !entities.contains(e)).forEach(entities::add);
}
```

Reply:
That won't work, unfortunately. The `getDependenciesStream` gives me only the one layer of dependencies while I need to collect all of them. Your example would allow collecting only that one layer because for entities added to the `entities` collection the `getDependenciesStream` is not called. It is because the iteration doesn't affect that collection. That's the reason I used that non-trivial approach.

Reply:
I think it is better to leave as is. All other solutions I came to are more complex and clunky. I would probably need to make comment more detailed to be more understandable.

Reply:
ok